### PR TITLE
PT-85: Set up user password repeat generation policy

### PR DESCRIPTION
**Pull request by eMazeika**
https://virtocommerce.atlassian.net/browse/PT-84

- the PasswordHistory feature is enabled/disabled and set by "IdentityOptions:Password:PasswordHistory" setting
- Custom IPasswordCheckService replaced by standard PasswordValidator<ApplicationUser> and CustomPasswordValidator
- Custom PasswordValidationResult replaced by standard IdentityResult
- Dedicated endpiont added for administrator to validate other users' passwords: ValidateUserPassword
- Password hash is recorded to history on:
    - user created
    - user password changed
    - user password reset

### CODE REVIEW

**Review by akak1977 — APPROVED** [review]

File: src/VirtoCommerce.Platform.Web/Security/CustomUserManager.cs [hunk 1]
```diff
@@ -22,19 +24,23 @@ public class CustomUserManager : AspNetUserManager<ApplicationUser>
         private readonly IEventPublisher _eventPublisher;
         private readonly IUserPasswordHasher _userPasswordHasher;
```
Comment (tatarincev) [inline src/VirtoCommerce.Platform.Web/Security/CustomUserManager.cs:25]: Could you make this IUserPasswordHasher obsolete with notes to use the new password history instead? Previously, IUserPasswordHasher was used for a single purpose only - to save password history.

File: src/VirtoCommerce.Platform.Security/Repositories/SecurityRepository.cs [hunk 2]
```diff
@@ -1,18 +1,37 @@
...
+using VirtoCommerce.Platform.Core.Security;
 using VirtoCommerce.Platform.Data.Infrastructure;
 using VirtoCommerce.Platform.Security.Model;
 
 namespace VirtoCommerce.Platform.Security.Repositories
 {
     public class SecurityRepository : DbContextRepositoryBase<SecurityDbContext>, ISecurityRepository
     {
-        public SecurityRepository(SecurityDbContext dbContext)
+        private readonly PasswordOptionsExtended _passwordOptions;
```
Comment (tatarincev) [inline src/VirtoCommerce.Platform.Security/Repositories/SecurityRepository.cs:14]: I think you should not inject PasswordOptionsExtended into this repository, please make the GetUserPasswordsHistoryAsync method more decoupled from the configuration options, better to receive the history count as an argument

File: src/VirtoCommerce.Platform.Web/Controllers/Api/SecurityController.cs [hunk 3]
```diff
@@ -539,9 +539,39 @@ public async Task<ActionResult> RequestPasswordReset(string loginOrEmail)
         [HttpPost]
         [Route("validatepassword")]
         [Authorize]
-        public async Task<ActionResult<PasswordValidationResult>> ValidatePassword([FromBody] string password)
+        public async Task<ActionResult<IdentityResult>> ValidatePassword([FromBody] string password)
```
Comment (tatarincev) [inline src/VirtoCommerce.Platform.Web/Controllers/Api/SecurityController.cs:542]: Mark this method as Obsolete with a comment to use   other version instead, and please and hide this method from SwaggerUI

**Review by tatarincev — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.Platform.Core/Security/PasswordOptionsExtended.cs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+namespace VirtoCommerce.Platform.Core.Security
+{
+    public class PasswordOptionsExtended
+    {
+        public int? PasswordHistory { get; set; }
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Core/Security/PasswordOptionsExtended.cs:5]: Please, add some info about this property.
``` csharp
/// <summary>
///  Very nice property
/// </summary>
```

File: src/VirtoCommerce.Platform.Security/CustomPasswordValidator.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+using Microsoft.AspNetCore.Identity;
+using VirtoCommerce.Platform.Core.Security;
+using VirtoCommerce.Platform.Security.Repositories;
+
+namespace VirtoCommerce.Platform.Security
+{
+    /// <summary>
+    /// Overriding https://github.com/dotnet/aspnetcore/blob/release/3.1/src/Identity/Extensions.Core/src/PasswordValidator.cs
+    /// </summary>
+    public class CustomPasswordValidator : PasswordValidator<ApplicationUser>
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Security/CustomPasswordValidator.cs:13]: Should we mark this class `sealed`?

File: src/VirtoCommerce.Platform.Security/Repositories/ISecurityRepository.cs [hunk 6]
```diff
@@ -1,11 +1,16 @@
...
 namespace VirtoCommerce.Platform.Security.Repositories
 {
     public interface ISecurityRepository : IRepository
     {
-        IQueryable<UserApiKeyEntity> UserApiKeys { get; }      
+        IQueryable<UserApiKeyEntity> UserApiKeys { get; }
+
+        IQueryable<UserPasswordHistoryEntity> UserPasswordsHistory { get; }
+
+        Task<UserPasswordHistoryEntity[]> GetUserPasswordsHistoryAsync(string userId);
```
Comment (Vectorfield4) [inline src/VirtoCommerce.Platform.Security/Repositories/ISecurityRepository.cs:14]: Any reason to use an array instead of IEnumerable<T>?

**Review by Vectorfield4 — APPROVED** [review]

File: tests/VirtoCommerce.Platform.Tests/Security/CustomPasswordValidatorTests.cs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        [InlineData("0987654321", true, true, true, true, 12, false, true, true, true, false, true)]
+        [InlineData("!@#$%^", true, true, true, false, 5, false, false, true, true, true, false)]
+        public async Task PasswordValidator_ValidatesByPasswordOptions(
+            string password,
+            bool requireUpperCase,
+            bool requireLowerCase,
+            bool requireDigits,
+            bool requireSpecialCharacters,
+            int minLength,
+            bool passwordIsValid,
```
Comment (Vectorfield4) [inline tests/VirtoCommerce.Platform.Tests/Security/CustomPasswordValidatorTests.cs:66]: Rename to expected.

File: tests/VirtoCommerce.Platform.Tests/Security/CustomPasswordValidatorTests.cs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        [InlineData("paS$word3", 4, false)]
+        [InlineData("paS$word4", 4, false)]
+        [InlineData("paS$word5", 4, true)]
+        [InlineData("paS$word6", 4, true)]
+        [InlineData("paS$word6", 6, false)]
+        [InlineData("paS$word6", 0, true)]
+        public async Task PasswordValidator_ValidatesByPasswordHistory(
+            string password,
+            int passwordRepeatHistory,
+            bool passwordIsValid)
```
Comment (Vectorfield4) [inline tests/VirtoCommerce.Platform.Tests/Security/CustomPasswordValidatorTests.cs:125]: Rename to expected.

File: tests/VirtoCommerce.Platform.Tests/Security/CustomPasswordValidatorTests.cs [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+using Moq;
+using VirtoCommerce.Platform.Core.Security;
+using VirtoCommerce.Platform.Security;
+using VirtoCommerce.Platform.Security.Model;
+using VirtoCommerce.Platform.Security.Repositories;
+using Xunit;
+
+namespace VirtoCommerce.Platform.Tests.Security
+{
+    public class CustomPasswordValidatorTests
```
Comment (Vectorfield4) [inline tests/VirtoCommerce.Platform.Tests/Security/CustomPasswordValidatorTests.cs:15]: Add comment above
```
// PT-1033: Refactor security mock helper and custom password validator
```

**Review by Vectorfield4 — APPROVED** [review]

File: src/VirtoCommerce.Platform.Security/CustomPasswordValidator.cs (hunk 5 above)
Comment (eMazeika) [reply src/VirtoCommerce.Platform.Security/CustomPasswordValidator.cs:13]: the base PasswordValidator is extendible. Why sealing?

Comment (Vectorfield4) [reply src/VirtoCommerce.Platform.Security/CustomPasswordValidator.cs:14]: CustomPasswordValidator is already is a special case. If the user needs an additional password validator he can implement it via IPasswordValidator\<T\>. By the way, where is `CustomPasswordValidator` registration? Maybe be I miss it, but I cant find it.

File: src/VirtoCommerce.Platform.Web/Controllers/Api/SecurityController.cs (hunk 3 above)
Comment (eMazeika) [reply src/VirtoCommerce.Platform.Web/Controllers/Api/SecurityController.cs:542]: the old PasswordValidationResult class was deleted, no way to keep it.

**Pull request merged**